stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -659,14 +659,19 @@ end
 # disambiguation methods: * of Diagonal and Adj/Trans AbsVec
 *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal) = Adjoint(map((t,s) -> t'*s, D.diag, parent(x)))
 *(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal) = Transpose(map((t,s) -> transpose(t)*s, D.diag, parent(x)))
-*(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-function dot(x::AbstractVector, D::Diagonal, y::AbstractVector)
-    mapreduce(t -> dot(t[1], t[2], t[3]), +, zip(x, D.diag, y))
+*(x::Adjoint{<:Any,<:AbstractVector},   D::Diagonal, y::AbstractVector) = _dot(x, D, y)
+*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) = _dot(x, D, y)
+dot(x::AbstractVector, D::Diagonal, y::AbstractVector) = _dot(x, D, y)
```

Comment:
This changes behavior. The idea is that `*` reduces only by one level, whereas `dot` acts recursively. For that reason, the naming `_dot` is unfortunate here, because it's not a helper function for `dot`, but for `*`. Now, `dot` probably deserves the same 0-length handling, but the `mapreduce` should continue to call `dot` as before.

Reply:
Apologies, I should have read the existing `dot` more carefully.